torch/onnx/_internal/fx/dynamo_exporter.py

```diff
@@ -1,25 +1,37 @@
+from __future__ import annotations
+
 import copy
-from typing import Optional
+import functools
+import inspect
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple, Type, Union
 
 import torch._dynamo
+import torch.fx
 import torch.onnx
-import torch.onnx._internal.fx.fx_exporter
+from torch.onnx._internal import _beartype
+from torch.onnx._internal.fx import fx_exporter
+from torch.utils import _pytree as pytree
 
 
-class DynamoOptimizeExporter(torch.onnx._internal.fx.fx_exporter.FXGraphModuleExporter):
+class DynamoOptimizeExporter(fx_exporter.FXGraphModuleExporter):
```

Comment:
My previous comments were based on the fact that the I/O adapters were meant as a helper for users, but now I see that you are also using it internally, right before the X graph extractors. Each FX extractor with a different set of "steps".

For the public use, I am inclined in having it as a separate class helper, away from `ExportOutput` or any other `torch.onnx` API. The helper wraps the `ExportOutput` and adapts input/output for the user and just give users the data in their desired format through `__call__`

However, for internal use as we see here, this was predicted by the original proposal through either 1) Pre-Process transformations or 2) [FXGraphExtractor](https://github.com/pytorch/pytorch/pull/98893). 

We could go with either 1) or 2), but I think 2) is probably the closest component to actually handle user data. `FXGraphExtractor` interface can be extended to add `input_adapter` and `output_adapter` methods that each FX API could implement independently. This way, the `Exporter` never needs to know about any of these IR differences

what do you think?

Reply:
Not only does different FX extractor may have different set of steps, the fx passes, and onnx passes in the future can too. I view them as "lego pieces" to be used in exporter. 

And as per docstring 

>        The actual formatting steps are associated with each individual export. It
>        depends on the PyTorch model, the particular set of model_args and model_kwargs
>        used for the export, and export options.

the formatting steps are as much "export output" as the model_proto. Hence the design.

Reply:
Good point, but I think this might be a different discussion.

The requirement for `ONNX Passes` is that they can either be executed 1) by the exporter after ONNX model is ready to be returned to the user or 2) by the user itself, outside the outside the exporter. In the current proposal, `ExportOutput` will not have the right formatter for scenario 2), as the transformation happened after `ExportOutput` was returned. An external data conversion helper would be needed, right? In fact, that might mean that part of the `ONNX Pass` API may have to "handle" data conversion for the users. Makes sense?

For the `FX Passes`, in general, they are internal and only valid before ONNX export. I believe they should not change the ONNX input/output schema. Otherwise, users are required to always use `ExportOutput`, because the ONNX input/output will not be "predictable" based on original input/output on pytorch eager mode. Makes sense?

Reply:
That's proving the point with this adapter right? With this design, for signature altering pass, it would be natural to emit adapter as their output along with the model. Again, the adapter serves the purpose to reduce guess work of "what happened to my inputs". If we design passes to operate on `ExportOutput`, then we'd expect the new `ExportOutput` would have internal adapter updated. If we design passes to operate on "Model IR", then original `ExportOutput` is an outdated snapshot.

Unless we restrict `FX Passes` to not change `FX` input/output schema, which is perhaps an interesting topic for discussion but clearly doesn't exist today, it will in turn affects ONNX. 

Bottom line is sure we can ditch this responsibility and have users figure things out as the old exporter did. But since we are pretty much at experimentation stage, I would like to see if we can export the signature adapting info too. In the torch-jit-paritybench benchmark on wild crawled projects and models, the highest percentage failure category for onnx export and validation is signature mismatch, higher than all long tail missing op symbolics combined.